presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java

```diff
@@ -437,7 +437,12 @@ public QueryId createQueryId()
         }
 
         // start the query in the background
-        resourceGroupManager.submit(preparedQuery.getStatement(), queryExecution, selectionContext, queryExecutor);
+        try {
+            resourceGroupManager.submit(preparedQuery.getStatement(), queryExecution, selectionContext, queryExecutor);
+        }
+        catch (Throwable e) {
```

Comment:
I'm just reading through the code and this isn't a proper review.  I hoped to learn a bit about how this area of the code works.

My question here is if there is there any way to accomplish this without catching `Throwable`?  For example, if `e` is an instance of some `Error` type which is serious and ought to cause the process to fail?  

Also, looking at the interface of `ResourceGroupManager`, I would expect this to be async--the `submit` method takes in an `Executor`, and when one sees this, it almost always means the `Executor` is used to parameterize the concurrency of the code that will be executed.  Perhaps this is a problem with the interface though.  But if the intent of the `submit` method on the `ResourceGroupManager` interface is to express that the action to be performed is to happen on the passed-in `Executor`, then I would expect a cleaner way to accomplish this would be to return a `ListenableFuture`, and we can add a callback on future error to fail the query (and don't do anything, for now, on success).

Reply:
Thanks, good point! Will think about this.

Reply:
Another option we discussed offline, but if the problem is that we have the potential to leak query ids, is a more appropriate place to put this handling inside the `createQuery` method, either in the catch or as a callback when the future fails.

Reply:
Hey @dain , @electrum advised that you're refactoring these files, when you have some time can you review this?

Reply:
There isn't a hard rule to this one.  Generally, when I catch `Throwable` and move it to another receiver (thread, future, etc), it is because the think being *tried* was a task on direct behalf of the it.  For example, when you submit a job to an executor, you get the exception from the task in the Future; otherwise there is nothing to handle the exception (in the case of `Executor.execute` it simply gets logged).  In this case, it is really important to fail the query and not hang it forever.  In some cases, depending on the context of the code, you may also want to rethrow the exception to make stop further processing, but it really comes down to use.

This actual fix is making over the bigger (known) bug in this code where the caller of this simply catches the exception and puts it into `queryCreationFuture.setException(e)`, and the processor of the this future does not handle failures correctly.  I looked at this a couple of weeks ago, and the full fix is pretty complex.

Reply:
Can we hoist this `failQuery` call to the parent method's catch block?  That way we don't preclude this future larger fix where we handle failures for the future (for example, if the exception is some subtype of `Error`, I doubt we'd want to suppress it).

Reply:
So then it makes more sense to catch RuntimeException instead of Error - since an error can't be recovered from (OutOfMemoryError, StackOverflowError, etc.).